src/mpid/ch4/shm/xpmem/xpmem_pre.h

```diff
@@ -27,6 +43,12 @@ typedef struct MPIDI_XPMEM_seg {
     MPIR_cc_t refcount;         /* reference count of this seg */
 } MPIDI_XPMEM_seg_t;
 
+typedef struct MPIDI_XPMEM_cnt {
+    MPIR_OBJECT_HEADER;
+    OPA_int_t pad;
+    OPA_int_t counter;
```

Comment:
Is `pad` unused? Can we delete it?

Reply:
no, we need to keep obj size at least 16 bytes due to `MPIR_Handle_common`.

Reply:
Then we should use union to ensure the max bytes required by `MPIR_Handle_common` are secured. The size of `void*` and  `OPA_int_t ` are all compiler/platform dependent, you can not assume `2*sizeof(OPA_int_t)` = `sizeof(void*)`. 

How about this? Note that I am not sure if **anonymous union** is allowed by MPICH as it is a C11 feature but likely we currently support only C99.
```
typedef union MPIDI_XPMEM_cnt {
    MPIR_Handle_common common; /* ensure sufficient bytes required for `MPIR_Handle_common` */
    struct {
        MPIR_OBJECT_HEADER;
        OPA_int_t counter;
    } obj;
} MPIDI_XPMEM_cnt_t

Reply:
OK, just confirmed that we cannot use **anonymous union**..Can you please also fix this code:
```
typedef struct {
    union {
#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
        MPIDI_SHM_ctrl_xpmem_send_lmt_rts_t xpmem_slmt_rts;
        MPIDI_SHM_ctrl_xpmem_send_lmt_cts_t xpmem_slmt_cts;
        MPIDI_SHM_ctrl_xpmem_send_lmt_send_fin_t xpmem_slmt_send_fin;
        MPIDI_SHM_ctrl_xpmem_send_lmt_recv_fin_t xpmem_slmt_recv_fin;
        MPIDI_SHM_ctrl_xpmem_send_lmt_cnt_free_t xpmem_slmt_cnt_free;
#endif
    };
} MPIDI_SHM_ctrl_hdr_t;
```
->
```
typedef union {
#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
        MPIDI_SHM_ctrl_xpmem_send_lmt_rts_t xpmem_slmt_rts;
        MPIDI_SHM_ctrl_xpmem_send_lmt_cts_t xpmem_slmt_cts;
        MPIDI_SHM_ctrl_xpmem_send_lmt_send_fin_t xpmem_slmt_send_fin;
        MPIDI_SHM_ctrl_xpmem_send_lmt_recv_fin_t xpmem_slmt_recv_fin;
        MPIDI_SHM_ctrl_xpmem_send_lmt_cnt_free_t xpmem_slmt_cnt_free;
#endif
} MPIDI_SHM_ctrl_hdr_t;
```

Reply:
you are right. fixing now.